src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs

```diff
@@ -447,6 +453,24 @@ internal string AdvanceThroughBalancedParentheticalExpression()
             return expressionText;
         }
 
+        internal void TakeSnapshot()

```

Comment:
Nit: It would be good to have TextPosition in the method names here to make it clear what is being snapshotted.